### ISSUE TITLE
MQTT connection fails silently after long period of time

### ISSUE THREAD

**Issue by jarredkenny** [aws/aws-iot-device-sdk-js-v2#108]
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-iot-device-sdk-js-v2/issues) and didn't find any solution
- [X] I'm using ATS data type endpoint: the endpoint should look like `<prefix>-ats.iot.<region>.amazonaws.com`

**Describe the bug**

**SDK version number**
- aws-crt `v1.3.2`
- aws-iot-device-sdk-v2 `v1.3.2`

**Platform/OS/Hardware/Device**
- Debian 10 (Buster)
- Node 12

**To Reproduce (observed behavior)**
- Create an MQTT connection using `mqtt.MqttClient`
- Connect to broker and subscribe to topics
- Wait approximately 18 hours (Difficult to confirm exact time)
- Device no longer receives messages on its subscribed topics

**Expected behavior**
- SDK should maintain its subscriptions indefinitely and the device should always receive messages on its subscribed topics.

**Logs/output**
After running my device over a 24 hour period I was able to capture DEBUG level logs from the SDK which contain a potentially relevant error. 

```
[DEBUG] [2020-11-22T15:51:46Z] [00007f309fea7d80] [dns] - id=0x1a0246293f30: Host resolution requested for a3e6d4yvq6d50l-ats.iot.us-east-2.amazonaws.com
[DEBUG] [2020-11-22T15:51:46Z] [00007f309fea7d80] [dns] - id=0x1a0246293f30: No cached entries found for a3e6d4yvq6d50l-ats.iot.us-east-2.amazonaws.com starting new resolver thread.
[DEBUG] [2020-11-22T15:51:46Z] [00007f3092fed700] [dns] - static: resolving host a3e6d4yvq6d50l-ats.iot.us-east-2.amazonaws.com
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
Error: libaws-c-mqtt: AWS_ERROR_MQTT_NOT_CONNECTED, The requested operation is invalid as the connection is not open.
    at /usr/share/ssn/player/resources/app/node_modules/aws-crt/dist/native/mqtt.js:318:39
    at new Promise (<anonymous>)
    at MqttClientConnection.<anonymous> (/usr/share/ssn/player/resources/app/node_modules/aws-crt/dist/native/mqtt.js:310:20)
    at Generator.next (<anonymous>)
    at /usr/share/ssn/player/resources/app/node_modules/aws-crt/dist/native/mqtt.js:31:71
    at new Promise (<anonymous>)
    at __awaiter (/usr/share/ssn/player/resources/app/node_modules/aws-crt/dist/native/mqtt.js:27:12)
    at MqttClientConnection.disconnect (/usr/share/ssn/player/resources/app/node_modules/aws-crt/dist/native/mqtt.js:309:16)
    at MessagingSystem.handleNetworkOffline (/usr/share/ssn/player/resources/app/lib/systems/messaging.js:137:29)
    at EventBus.emit (events.js:228:7)
[DEBUG] [2020-11-22T15:51:53Z] [00007f309fea7d80] [mqtt-client] - id=0x1a0246c31a80: user called disconnect.
[ERROR] [2020-11-22T15:51:53Z] [00007f309fea7d80] [mqtt-client] - id=0x1a0246c31a80: Connection is not open, and may not be closed
[ERROR] [2020-11-22T15:52:06Z] [00007f3092fed700] [dns] - static: getaddrinfo failed with error_code -3
[ERROR] [2020-11-22T15:52:06Z] [00007f3092fed700] [channel-bootstrap] - id=0x1a0246702de0: dns resolution failed, or all socket connections to the endpoint failed.
[DEBUG] [2020-11-22T15:52:06Z] [00007f3092fed700] [channel-bootstrap] - id=0x1a0246702de0: releasing bootstrap reference
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
Failed to connect: aws-c-io: AWS_IO_DNS_QUERY_FAILED, A query to dns failed to resolve.
```

Full log here:
[aws-sdk.stdout.log](https://github.com/aws/aws-iot-device-sdk-js-v2/files/5591889/aws-sdk.stdout.log)

**Comment by bretambrose** [comment]
Just skimming your snippet and then the log, it looks like you manually disconnect a connection about 10 seconds prior (15:51:36) and then try to create a new one.  The first connection is gone and the second connection does not appear to ever successfully complete because you're in the middle of an outage.  After that I see another call to disconnect on that never-completed connection and then a bunch of publish attempts which aren't going to succeed.

So I think there's multiple issues going on here, some potentially ours, some potentially yours.

On our side I think we should investigate:
1. Are we properly surfacing initial connection failures to the user?  Initial connection failures will *not* be retried.
1. Given that we don't retry connections that initially fail, should operations like publish and subscribe be treated differently (exceptions vs queuing)?

On your side, consider:
1. Is it necessary to manually disconnect and make a new connection?  In an outage, a previously-established connection will attempt to reconnect indefinitely, so it's not clear that this complication is the right choice.
1. Are you handling an initial connection failure (will not reconnect) properly?

**Comment by jarredkenny** [comment]
>  it looks like you manually disconnect a connection about 10 seconds prior (15:51:36) and then try to create a new one.

Reviewing our application logs, I don't see an offline event that lines up with that time period. In our current implementation, we only call `disconnect()` on the MQTT connection when we've detected we are offline.

> Is it necessary to manually disconnect and make a new connection? In an outage, a previously-established connection will attempt to reconnect indefinitely, so it's not clear that this complication is the right choice.

We've written the code on our side to do this because the SDK was not reconnecting automatically when the network was disconnected and reconnected after a period of time. Our current implementation polls the devices network interfaces, disconnects the AWS IOT connection when we are offline, and creates a new connection when we are online again.

> Are you handling an initial connection failure (will not reconnect) properly?

Because we only attempt a connection when we know we have a route to the internet, we have not yet run into this case.


It does sound like we don't need to be manually disconnecting and reconnected when we've detected our network status has changed. However, this definitely was not working and is why we went with this approach. I will try to revert some of our code to let the SDK manage its connection state and see what I see.

**Comment by jarredkenny** [comment]
I've reworked some of our code for some initial testing and it does appear that the SDK (v1.3.2) will reconnect automatically when an internet connection is restored to the device. We weren't see this work correctly on earlier versions, which is why we decided to create a new MQTT connection when we determined we were back online.

With this now working, the SDK will automatically reconnect and receive any messages that were missed when the device returns online as long as `with_clean_session` is `false` in our MQTT connection configuration. When we were disconnecting and and creating a new connection oursleves we were using `with_clean_session` `true` and manually subscribing when the connection was made.

Now that reconnection seems to be working, we will refactor our code to make use of it. However, this does leave me with a couple of questions: 

1. Will the SDK attempt to reconnect to the broker indefinitely, or is there a timeout at which the SDK no longer attempts to restore the connection.

2. We rely heavily on `subscribeToNamedShadowDeltaUpdatedEvents` to alert our devices to configuration changes written to a named device shadow. Will the device always receive all named shadow deltas if offline for an extended period of time? Will we receive a delta for each write to the named shadow that occurred while the device was offline, or will we only receive one delta accounting for all shadow writes that occurred while the device was offline? 

When reconnection was not working, our code created a new connection when our device came back online, and used `subscribeToGetNamedShadowAccepted`/`publishGetNamedShadow` to fetch the latest device shadow state and resolve deltas as needed.

**Comment by jarredkenny** [comment]
We've had some time to test things with our new approach of letting the IOT SDK handle reconnection but unfortunately are still facing some of the same issues.

As I said above, I've updated things on our end to only handle connection errors throw on initial connection. After the initial connection is made, I leave it up to the SDK to handle reconnection. 

The scenario is which this fails consistently for us is when we physically powered down the wireless access point the device is connected to for a period of time (2 hours in our testing). This seems to be a different scenario then simply disconnecting the device from the network for a period of time. When we disconnect from the device and reconnect, things seems to work. However, when powerering down our AP for a period of time we see this failure.

When the network is brought back only the device appears to reconnect to the broker (confirmed via reserved presence topics) but it does not receive messages on its subscribed topics.

In an attempt to make this work, I tried having our code resubscribe to its MQTT topics when we detect we are back online, but this has not worked for us either .

Again, we have `with_clean_session: false` in our configuration, so I would expect the MQTT subscriptions to resume in either case. 

Here are debug logs from the SDK capturing the time period in which the network goes down and back up. When the network comes back online the SDK appears to reconnect, but will never receive messages on its MQTT subscriptions.  

 
[iot.stdout.log](https://github.com/aws/aws-iot-device-sdk-js-v2/files/5652912/ssn-content-player.stdout.log)
